src/core/Akka/Serialization/Serialization.cs

```diff
@@ -120,6 +179,47 @@ public Serialization(ExtendedActorSystem system)
             }
         }
 
+        private Information SerializationInfo => System.Provider.SerializationInformation;
+
+        /// <summary>
+        /// Performs the requested serialization function while also setting
+        /// the <see cref="CurrentTransportInformation"/> based on available data
+        /// from the <see cref="ActorSystem"/>. Useful when serializing <see cref="IActorRef"/>s.
+        /// </summary>
+        /// <typeparam name="T">The type of message being serialized.</typeparam>
+        /// <param name="system">The <see cref="ActorSystem"/> performing serialization.</param>
+        /// <param name="action">The serialization function.</param>
+        /// <returns>The serialization output.</returns>
+        public static T WithTransport<T>(ExtendedActorSystem system, Func<T> action)
+        {
+            var info = system.Provider.SerializationInformation;
+            if (CurrentTransportInformation == info)
+            {
+                // already set
+                return action();
+            }
+            else
+            {
+                CurrentTransportInformation = info;
```

Comment:
Previous transport information is never restored after calling this method - is that how it's supposed to work?

Reply:
Yeah that might be a mistake - let me double check

Reply:
So I think this is some weird scala syntax:

```scala
val info = system.provider.serializationInformation
    if (Serialization.currentTransportInformation.value eq info)
      f() // already set
    else
      Serialization.currentTransportInformation.withValue(info) {
        f()
      }
  }
```

I'm going to interpret the scoping operating there like a `using` block - the `currentTransportInformation` only has that value for the duration of that call. Therefore, this is probably a bug in the way I translated the code. I'll change it so it unsets the `CurrentTransportInformation` at the end. Good catch.

Reply:
More explanation on `DynamicVariable` in Scala: https://stackoverflow.com/questions/5116352/when-we-should-use-scala-util-dynamicvariable

Reply:
This has been resolved.